### ISSUE TITLE
Workflow loses its `workflows.argoproj.io/phase` label and is considered as unreconciled after it's reapplied with kubectl

### ISSUE THREAD

**Issue by pavelsmolensky** [argoproj/argo-workflows#12601]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issue exists when I tested with `:latest`
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

1. Apply the following workflow definition. It's important that it contains labels in metadata
`kubectl -n sandbox apply -f wf-1.yaml`
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: test-wf
  labels:
    xxx: aaa
spec:
  entrypoint: test
  templates:
  - name: test
    container:
      image: docker/whalesay
      command: [ cowsay ]
      args: [ "hello" ]
      resources:
        limits:
          memory: 32Mi
          cpu: 100m
```
2. Wait until the workflow run completes
2a. Note that besides label `xxx: aaa` 2 new labels appeared:
`kubectl -n sandbox get workflow test-wf -oyaml`

```
...
labels:
  workflows.argoproj.io/completed: "true"
  workflows.argoproj.io/phase: Succeeded
  xxx: aaa
...
```

3. Reapply the same workflow definition, changing its args and **removing labels**:

`kubectl -n sandbox apply -f wf-2.yaml`
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: test-wf
spec:
  entrypoint: test
  templates:
  - name: test
    container:
      image: docker/whalesay
      command: [ cowsay ]
      args: [ "bye" ]
      resources:
        limits:
          memory: 32Mi
          cpu: 100m
```
Result:
No new run of changed workflow is triggered, and applied workflow definition loses `workflows.argoproj.io/phase: Succeeded` label, resulting in the following label left:
```
labels: 
  workflows.argoproj.io/completed: "true"
```

Because of that healthcheck method of workflow controller starts detecting this workflow as unreconciled: https://github.com/argoproj/argo-workflows/blob/main/workflow/controller/healthz.go#L47-L57
It leads to controller's pod health probe failing and causes regular controller pod restarts

### Version

v3.5.4

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
see above
```

### Logs from the workflow controller

```text
time="2024-01-31T17:31:03.967Z" level=info msg="Processing workflow" Phase= ResourceVersion=118463669 namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:03.970Z" level=warning msg="Non-transient error: configmaps \"artifact-repositories\" not found"
time="2024-01-31T17:31:03.970Z" level=info msg="resolved artifact repository" artifactRepositoryRef=default-artifact-repository
time="2024-01-31T17:31:03.970Z" level=info msg="Task-result reconciliation" namespace=sandbox numObjs=0 workflow=test-wf
time="2024-01-31T17:31:03.970Z" level=info msg="Updated phase  -> Running" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:03.970Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:03.970Z" level=info msg="was unable to obtain node for , letting display name to be nodeName" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:03.970Z" level=info msg="Pod node test-wf initialized Pending" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:04.079Z" level=info msg="Created pod: test-wf (test-wf)" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:04.079Z" level=info msg="TaskSet Reconciliation" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:04.079Z" level=info msg=reconcileAgentPod namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:04.079Z" level=info msg="Workflow to be dehydrated" Workflow Size=1255
time="2024-01-31T17:31:04.091Z" level=info msg="Workflow update successful" namespace=sandbox phase=Running resourceVersion=118463674 workflow=test-wf
time="2024-01-31T17:31:14.079Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=118463674 namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:14.079Z" level=info msg="Task-result reconciliation" namespace=sandbox numObjs=1 workflow=test-wf
time="2024-01-31T17:31:14.079Z" level=info msg="task-result changed" namespace=sandbox nodeID=test-wf workflow=test-wf
time="2024-01-31T17:31:14.079Z" level=info msg="node changed" namespace=sandbox new.message= new.phase=Succeeded new.progress=0/1 nodeID=test-wf old.message= old.phase=Pending old.progress=0/1 workflow=test-wf
time="2024-01-31T17:31:14.079Z" level=info msg="TaskSet Reconciliation" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:14.079Z" level=info msg=reconcileAgentPod namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:14.079Z" level=info msg="Updated phase Running -> Succeeded" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:14.079Z" level=info msg="Marking workflow completed" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:14.079Z" level=info msg="Workflow to be dehydrated" Workflow Size=1657
time="2024-01-31T17:31:14.084Z" level=info msg="cleaning up pod" action=deletePod key=sandbox/test-wf-1340600742-agent/deletePod
time="2024-01-31T17:31:14.093Z" level=info msg="Workflow update successful" namespace=sandbox phase=Succeeded resourceVersion=118463843 workflow=test-wf
time="2024-01-31T17:31:14.108Z" level=info msg="cleaning up pod" action=labelPodCompleted key=sandbox/test-wf/labelPodCompleted
time="2024-01-31T17:31:35.677Z" level=info msg="Processing workflow" Phase=Succeeded ResourceVersion=118464139 namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:35.677Z" level=info msg="Task-result reconciliation" namespace=sandbox numObjs=0 workflow=test-wf
time="2024-01-31T17:31:35.677Z" level=info msg="TaskSet Reconciliation" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:35.677Z" level=info msg=reconcileAgentPod namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:35.677Z" level=info msg="Marking workflow completed" namespace=sandbox workflow=test-wf
time="2024-01-31T17:31:35.677Z" level=info msg="Workflow to be dehydrated" Workflow Size=1454
time="2024-01-31T17:31:35.683Z" level=info msg="cleaning up pod" action=deletePod key=sandbox/test-wf-1340600742-agent/deletePod
time="2024-01-31T17:31:35.690Z" level=info msg="Workflow update successful" namespace=sandbox phase=Succeeded resourceVersion=118464140 workflow=test-wf
time="2024-01-31T17:34:02.522Z" level=info msg="Alloc=16953 TotalAlloc=184755 Sys=51557 NumGC=24 Goroutines=177"
time="2024-01-31T17:34:02.690Z" level=info msg="Processing sandbox/test-cron-wf" cronWorkflow=sandbox/test-cron-wf
time="2024-01-31T17:34:02.740Z" level=info msg="CronWorkflow sandbox/test-cron-wf added" cronWorkflow=sandbox/test-cron-wf
time="2024-01-31T17:35:00.000Z" level=info msg="Running test-cron-wf" namespace=sandbox workflow=test-cron-wf
time="2024-01-31T17:35:00.002Z" level=info msg="test-cron-wf is suspended, skipping execution" namespace=sandbox workflow=test-cron-wf
time="2024-01-31T17:37:02.291Z" level=info msg=healthz age=5m0s err="workflow never reconciled: test-wf" instanceID= labelSelector="!workflows.argoproj.io/phase,!workflows.argoproj.io/controller-instanceid" managedNamespace=
time="2024-01-31T17:38:02.292Z" level=info msg=healthz age=5m0s err="workflow never reconciled: test-wf" instanceID= labelSelector="!workflows.argoproj.io/phase,!workflows.argoproj.io/controller-instanceid" managedNamespace=
time="2024-01-31T17:39:02.293Z" level=info msg=healthz age=5m0s err="workflow never reconciled: test-wf" instanceID= labelSelector="!workflows.argoproj.io/phase,!workflows.argoproj.io/controller-instanceid" managedNamespace=
```

### Logs from in your workflow's wait container

```text
time="2024-01-31T17:31:05.675Z" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
time="2024-01-31T17:31:05.675Z" level=info msg="Executor initialized" deadline="0001-01-01 00:00:00 +0000 UTC" includeScriptOutput=false namespace=sandbox podName=test-wf templateName=test version="&Version{Version:v3.5.4,BuildDate:2024-01-14T05:30:37Z,GitCommit:960af331a8c0a3f2e263c8b90f1daf4303816ba8,GitTag:v3.5.4,GitTreeState:clean,GoVersion:go1.21.6,Compiler:gc,Platform:linux/amd64,}"
time="2024-01-31T17:31:05.689Z" level=info msg="Starting deadline monitor"
time="2024-01-31T17:31:07.691Z" level=info msg="Main container completed" error="<nil>"
time="2024-01-31T17:31:07.691Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2024-01-31T17:31:07.691Z" level=info msg="No output parameters"
time="2024-01-31T17:31:07.691Z" level=info msg="No output artifacts"
time="2024-01-31T17:31:07.713Z" level=info msg="Alloc=7870 TotalAlloc=13845 Sys=30565 NumGC=4 Goroutines=8"
time="2024-01-31T17:31:07.724Z" level=info msg="stopping progress monitor (context done)" error="context canceled"
time="2024-01-31T17:31:07.724Z" level=info msg="Deadline monitor stopped"
```

**Comment by Joibel** [comment]
I may have an idea what's going on here, I'll have a look into it tomorrow.

**Comment by agilgur5** [comment]
> I may have an idea what's going on here, I'll have a look into it tomorrow.

Related to #12198 I assume? Since this is intentionally submitting 2 workflows with the same name in quick succession. This logic might need to be changed to use UID?

**Comment by Joibel** [comment]
This is already fixed by #12537. This isn't available in a release yet. If it's a problem revert to v3.5.2.

#12198 was one of my suspicions, but that shouldn't have allowed this to be processed at all, and this scenario was one I had envisaged and tested.

In 3.4.16 the label `workflows.argoproj.io/completed: "true"` doesn't get reapplied. In 3.5.2 it does get reapplied because of the `status`, but the `workflows.argoproj.io/phase: Succeeded` also gets reapplied, so everything is happy.

3.5.4 reapplies the `completed` label, but not the `phase` label which is the cause of the healthz failure, but #12537 fixes this.

@pavelsmolensky - thanks for the detailed bug report, it made reproducing the problem easy.

**Comment by pavelsmolensky** [comment]
@Joibel thanks for the prompt feedback!
